[1/X] Store user data for username, email, and whether email should be shown for comments in persistent storage.

nikhitashanker opened the pull request:
-Add form to save users' user name and whether they want their email to be displayed for comments in persistent storage.
![Screenshot 2020-07-10 at 10 58 33 AM](https://user-images.githubusercontent.com/12983270/87184260-8c22d080-c29c-11ea-9df7-294a535612c2.png)

-Add username and email to the comment from persistent storage.
![Screenshot 2020-07-10 at 11 01 08 AM](https://user-images.githubusercontent.com/12983270/87184450-e02db500-c29c-11ea-90eb-7fad95b32402.png)

yicyao commented on portfolio/src/main/java/com/google/sps/servlets/FormHandlerServlet.java:
```diff
@@ -33,9 +37,8 @@
-  private static final String EMAIL = "commenter-email";
   private static final String IMAGE = "image";
 
   @Override
```
add javadoc to doPost method

nikhitashanker replied:
iirc, this goes under the self-explanatory exception to javadoc.

yicyao commented on portfolio/src/main/java/com/google/sps/servlets/UserInfoServlet.java:
```diff
@@ -0,0 +1,74 @@
+  private static final UserService userService = UserServiceFactory.getUserService();
+  private static final DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
+
+  @Override
+  public void doGet(HttpServletRequest request, HttpServletResponse response) throws IOException {
+    if (userService.isUserLoggedIn()) {
+      // Send user info as JSON response.
+      response.setContentType("application/json;");
+      UserInfo userInfo = UserInfoUtils.getUserInfo(userService.getCurrentUser().getUserId());
+      response.getWriter().println(CommonUtils.convertToJson(userInfo));
+    } else {
+      response.sendError(HttpServletResponse.SC_UNAUTHORIZED, "User is not logged in.");
+    }
+  }
+
+  @Override
+  public void doPost(HttpServletRequest request, HttpServletResponse response) throws IOException {
+    if (!userService.isUserLoggedIn()) {
+      response.sendRedirect("/login");
+      return;
+    }
+
```
add javadocs / comment what the functions do

nikhitashanker replied:
iirc, this goes under the self-explanatory exception to javadoc.

dash102 commented on portfolio/src/main/java/com/google/sps/servlets/UserInfoServlet.java:
```diff
@@ -0,0 +1,74 @@
+    User currentUser = userService.getCurrentUser();
+    String email = currentUser.getEmail();
+    String id = currentUser.getUserId();
+    boolean showEmail = request.getParameter(SHOW_EMAIL) == null ? false : true;
```
Can be shortened to not require ternary operators, afaik

nikhitashanker replied:
Done - Yes, I think that is correct.

dash102 commented on portfolio/src/main/webapp/script.js:
```diff
@@ -197,27 +196,39 @@ function showCommentFormAndLoginPrompt() {
 }
 
-function showLoginOrLogoutForm(isLoggedIn) {
+function showLoginOrLogoutForm(isLoggedIn, email) {
```
The new `email` parameter seems extraneous, since it's not used in the function and not an argument when it's called above

nikhitashanker replied:
Done - removed extraneous parameter email.

dash102 commented on portfolio/src/main/webapp/style.css:
```diff
@@ -2,6 +2,18 @@ body {
 }
+
+button {
+  background-color: pink; /* Green */
+  border: none;
+  color: black;
+  padding: 0.5rem 1rem;
+  text-align: center;
+  text-decoration: none;
+  display: inline-block;
+  font-size: 1rem;
```
(small: alphabetized attributes per block)

nikhitashanker replied:
Done.

hlz03 commented:
> Currently, each time a user enters a comment the user must enter their username and their email is always displayed as the email they are logged in with. In this change, the user can use a form to save their user name and whether they want their email to be displayed in persistent storage. Consequently, whenever they post comments this username is included and the email is included depending on whether they opted for their email to be displayed. In addition, we add a greeting that tells the user the email of the user that is signed in. Some styling was also added.

Usually we don't write description paragraph, use a bullet point list should be enough, not need to explain the whole logic (actually no one would look at these). You can take Yichen's PR as a good example.

hlz03 commented on portfolio/src/main/java/com/google/sps/servlets/FormHandlerServlet.java:
```diff
@@ -51,13 +54,25 @@ public void doPost(HttpServletRequest request, HttpServletResponse response) thr
+    // Set commenter email to email from UserService if they opted to show their email and Unknown
+    // otherwise.
+    UserInfo userInfo = UserInfoUtils.getUserInfo(userService.getCurrentUser().getUserId());
+    String commenterEmail;
+    if (userInfo == null) {
+      commenterEmail = "Unknown";
+    } else {
+      String userEmail = userInfo.getEmail();
+      commenterEmail = userInfo.getShowEmail() ? userEmail : "Unknown";
+    }
```
can be a separate method, here can just be commenterEmail = getCommenterEmail

nit: the logic can be simpler:

if (userInfo != null && userInfo.GetShowEmail()) {
return useremail.
}
return "Unknown"

nikhitashanker replied:
Done - fixed logic.

nikhitashanker commented:
> > Currently, each time a user enters a comment the user must enter their username and their email is always displayed as the email they are logged in with. In this change, the user can use a form to save their user name and whether they want their email to be displayed in persistent storage. Consequently, whenever they post comments this username is included and the email is included depending on whether they opted for their email to be displayed. In addition, we add a greeting that tells the user the email of the user that is signed in. Some styling was also added.
> 
> Usually we don't write description paragraph, use a bullet point list should be enough, not need to explain the whole logic (actually no one would look at these). You can take Yichen's PR as a good example.

Noted - I have changed the description. I will follow this format for future PRs.

yicyao commented on portfolio/src/main/webapp/comments.html:
```diff
@@ -0,0 +1,54 @@
+      href="https://fonts.googleapis.com/css2?family=Open+Sans&display=swap"
+      rel="stylesheet"
+    />
+
```
delete space?

nikhitashanker replied:
Done - thanks.

nikhitashanker commented:
Adding fix for removing network requests in a separate PR.

liangqigong commented on portfolio/src/main/java/com/google/sps/servlets/FormHandlerServlet.java:
```diff
@@ -51,13 +54,18 @@ public void doPost(HttpServletRequest request, HttpServletResponse response) thr
+    String commenterEmail = getCommenterEmail(userInfo);
+
+    // Set commenter name to username if user provided one and Anonymous otherwise.
+    String commenterName = userInfo == null ? "Anonymous" : userInfo.getUsername();
```
Make anonymous a const

nikhitashanker replied:
Done.

liangqigong commented on portfolio/src/main/java/com/google/sps/servlets/FormHandlerServlet.java:
```diff
@@ -104,4 +112,13 @@ private static String getUploadedFileUrl(HttpServletRequest request, String form
+    if (userInfo != null && userInfo.getShowEmail()) {
+      return userInfo.getEmail();
+    }
+    return "Unknown";
```
Use const for unknown as well

nikhitashanker replied:
Done.

liangqigong commented on portfolio/src/main/java/com/google/sps/servlets/UserInfoServlet.java:
```diff
@@ -0,0 +1,74 @@
+
+  @Override
+  public void doGet(HttpServletRequest request, HttpServletResponse response) throws IOException {
+    if (userService.isUserLoggedIn()) {
```
Usually we prefer to check if there is error and early return. So here, it would be better to check if user is not logged in, and return it so.

Just find one related document from Google.
https://www.itamarweiss.com/personal/2018/02/28/return-early-pattern.html

hlz03 commented on portfolio/src/main/java/com/google/sps/data/UserInfo.java:
```diff
@@ -0,0 +1,25 @@
+package com.google.sps.data;
+
+public class UserInfo {
+  private String email;
+  private boolean showEmail;
+  private String username;
```
Can be private final if only initialize once.

nikhitashanker replied:
Done.
